src/compiler/output-targets/dist-custom-elements/index.ts

```diff
@@ -233,49 +231,76 @@ export const addCustomElementInputs = (
-            }
-        });
-    }
-};`;
-  }
+  // Generate the contents of the entry file to be created by the bundler
+  bundleOpts.loader!['\0core'] = generateEntryPoint(outputTarget, indexImports, indexExports, exportNames);
 };
 
 /**
  * Generate the entrypoint (`index.ts` file) contents for the `dist-custom-elements` output target
  * @param outputTarget the output target's configuration
+ * @param cmpImports The import declarations for local component modules.
+ * @param cmpExports The export declarations for local component modules.
+ * @param cmpNames The exported component names (could be aliased) from local component modules.
  * @returns the stringified contents to be placed in the entrypoint
  */
-export const generateEntryPoint = (outputTarget: d.OutputTargetDistCustomElements): string => {
-  const imp: string[] = [];
+export const generateEntryPoint = (
+  outputTarget: d.OutputTargetDistCustomElements,
+  cmpImports: string[] = [],
+  cmpExports: string[] = [],
+  cmpNames: string[] = []
+): string => {
+  const body: string[] = [];
+  const imports: string[] = [];
+  const exports: string[] = [];
 
-  imp.push(
+  // Exports that are always present
+  exports.push(
     `export { setAssetPath, setPlatformOptions } from '${STENCIL_INTERNAL_CLIENT_ID}';`,
     `export * from '${USER_INDEX_ENTRY_ID}';`
   );
 
+  // Content related to global scripts
   if (outputTarget.includeGlobalScripts !== false) {
-    imp.push(`import { globalScripts } from '${STENCIL_APP_GLOBALS_ID}';`, `globalScripts();`);
+    imports.push(`import { globalScripts } from '${STENCIL_APP_GLOBALS_ID}';`);
+    body.push(`globalScripts();`);
   }
 
-  return imp.join('\n') + '\n';
+  // Content related to the `bundle` export behavior
+  if (outputTarget.customElementsExportBehavior === 'bundle') {
+    imports.push(...cmpImports);
+    body.push(
+      'export const defineCustomElements = (opts) => {',
+      "    if (typeof customElements !== 'undefined') {",
+      '        [',
+      ...cmpNames.map((cmp) => `            ${cmp},`),
+      '        ].forEach(cmp => {',
+      '            if (!customElements.get(cmp.is)) {',
+      '                customElements.define(cmp.is, cmp, opts);',
+      '            }',
+      '        });',
+      '    }',
+      '};'
+    );
+  }
+
+  // Content related to the `single-export-module` export behavior
+  if (outputTarget.customElementsExportBehavior === 'single-export-module') {
+    exports.push(...cmpExports);
+  }
+
+  // Generate the contents of the file based on the parts
+  // defined above. This keeps the file structure consistent as
+  // new export behaviors may be added
+  let content = '';
+
+  // Add imports to file content
+  content += imports.length ? imports.join('\n') + '\n' : '';
+  // Add exports to file content
+  content += exports.length ? exports.join('\n') + '\n' : '';
+  // Add body to file content
+  content += body.length ? body.join('\n') + '\n' : '';
```

Comment:
wondering if we should put an extra newline before the body content? just so we get something like

```ts
import { globalScripts } from '${STENCIL_APP_GLOBALS_ID}';
export { setAssetPath, setPlatformOptions } from '${STENCIL_INTERNAL_CLIENT_ID}';
export * from '${USER_INDEX_ENTRY_ID}';

globalScripts();
```

just for a little readability improvement